src/include/stat.h

```diff
@@ -6,35 +6,169 @@
+ * a heuristic for the number of CPUs. Unfortunately, However, inside WT we do
+ * not know when the application creates its threads.
+ *
+ * Our solution is to simply use a fixed number of slots that roughly
+ * approximates the largest number of cores we expect to see on the machine
+ * where WT is used. All we want is to avoid all threads writing the same
+ * variable at the same time, and that solution does the trick.
+ *
+ */
+#define	WT_COUNTER_SLOTS 24
+
+/* This is a cache-line-padded counter value. 
+ * It is here to effectively support array counters.
+ * Padding is needed, otherwise cache coherency messages
+ * will be triggered because of false sharing.
+ *
+ * The actual counter value v must be signed, because it is possible
+ * that one thread incremented the counter in its own slot, and then
+ * another thread decremented the same counter in another slot,
+ * which was initially zero. We need to allow the value is the second
+ * thread's slot to be negative. When the values are aggregated, we get
+ * the correct total value.
+ */
+typedef struct{
+	int64_t v;
+	char padding[WT_CACHE_LINE_ALIGNMENT-sizeof(uint64_t)];
+} padded_counter_t;
+
 struct __wt_stats {
 	const char	*desc;				/* text description */
-	uint64_t	 v;				/* 64-bit value */
+//  	uint64_t	 v;				/* 64-bit value */
+	padded_counter_t array_v[WT_COUNTER_SLOTS];     /* Array counter */
 };
 
 /*
+ * This macro determines the slot id for the array of counters.
+ *
+ * Ideally, we want a slot per CPU, and we want each thread to index the slot
+ * corresponding to the CPU it runs on. Unfortunately, getting the id of the
+ * current CPU is difficult. Some operating systems provide a system call for
+ * that, but not all. Further, invoking a system call every time we need to
+ * increment a stats counter is expensive. Another option would be to use the
+ * rdtscp instruction, but it's x86 specific, has a ~50 cycle overhead and is
+ * not supported on some older processors.
+ *
+ * Our second-best option is to use the thread ID. Unfortunately, there is no
+ * portable way to obtain a thread ID that is a small-enough number that could
+ * be used as an array index.
+ *
+ * Our solution is to use the session ID, because there is a session per thread
+ * and the id itself is a small monotonically increasing number, so this results
+ * in threads writing into different array slots, which is ultimately what we
+ * want.
+ */
+#define	WT_STATS_SLOT_ID	(session->id) % WT_COUNTER_SLOTS
+
+/* Set all the values in the array counter slots to zero. We are
+ * doing more work than needed by using memset, because we are setting
+ * both the values and the padding. However, resetting the counters is
+ * not a common case operation, so we use memset for compactness.
+ */
+#define	WT_STAT_ALL_RESET(stats, fld)  do {				\
+		memset((stats)->fld.array_v, 0,                         \
+		       sizeof(padded_counter_t) * WT_COUNTER_SLOTS);	\
+} while (0)
```

Comment:
@keithbostic: Haven't measured, because did not realize that resetting counters is a common operation (based on code reading). I can certainly measure. Could you recommend tests that perform lots of counter resetting? Are there any in the tree? My test that suffers from scalability loss due to stats counters, which I've been primarily using to test this branch, does not reset counters, so it's not the right test to evaluate this effect.